doc/memos/rdm-draft-weiss-exposing-api-firmware-guidelines.md

```diff
@@ -0,0 +1,155 @@
+- rdm-draft-weiss-exposing-api-firmware-guidelines.md
+- Title: Guidelines for Write Firmware to Expose RIOT APIs
+- Author: Kevin Weiss
+- Status: In-progress
+- Type: Process
+- Date: December 2018
+
+## Abstract
+
+In order to improve the testing and interface tools for RIOT a standard way of
+exposing API calls to a higher level should be defined. This document describes
+a set of guidelines indicating semantics and basic design on the firmware level
+that allow interfaces and tools to be build upon.
+
+## Status
+
+This document is currently under open discussion. Its content is licensed with
+a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the RFC2119 terminology and the following acronyms:
+API: Application Programming Interface
+cmd: Command
+msg: Message
+
+# 1. Introduction
+The guidelines are used indicate the semantics and design when writing
+firmware to expose APIs through the shell.
+
+_NOTE: It may not be applicable for every API call_
+
+# 2. Test Firmware Design
+
+The following guidelines are used when writing the C program to be flashed to
+the board.
+
+## 2.1. Test Firmware Declarations
+
+- Tests should interact through a shell
+- Commands should be synchronous
+- Tests should have a `get_metadata` command
+- The `get_metadata` command should contain at least board name, and
+application name
+- Tests should have a `echo <random 8 hex digits>` command
```

Comment:
Yup, I will try to implement something, I suppose it would be like the ps/reboot/ect. where you can just include it so there will be no code duplication.